File: app/src/main/java/hsk/practice/myvoca/VocaViewModel.java
package hsk.practice.myvoca;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;

import java.util.List;

import Database.Vocabulary;
import Database.source.VocaRepository;


public class VocaViewModel extends ViewModel {

    private LiveData<List<Vocabulary>> allVocabularies;
    private Vocabulary vocabulary;
    private VocaRepository vocaRepo;

    public VocaViewModel() {
        this.vocaRepo = VocaRepository.getInstance();
        loadVocabularies();
    }

    public LiveData<List<Vocabulary>> getAllVocabulary() {
        if (allVocabularies == null || allVocabularies.getValue() == null) {
            loadVocabularies();
        }
        return allVocabularies;
    }

    public LiveData<Integer> getVocabularyCount() {
        final MutableLiveData<Integer> result = new MutableLiveData<>();
        if (allVocabularies == null || allVocabularies.getValue() == null) {
            loadVocabularies();
            allVocabularies.observeForever(new Observer<List<Vocabulary>>() {
                @Override
                public void onChanged(List<Vocabulary> vocabularies) {
                    result.setValue(vocabularies.size());
                    allVocabularies.removeObserver(this);
                }
            });
        } else {
            result.setValue(allVocabularies.getValue().size());
        }
        return result;
    }

    public void deleteVocabulary(Vocabulary... vocabularies) {
        vocaRepo.deleteVocabularies(vocabularies);
    }

    public LiveData<List<Vocabulary>> getVocabulary(String query) {
        return vocaRepo.getVocabulary(query);
    }

    public void insertVocabulary(Vocabulary... vocabularies) {
        vocaRepo.insertVocabulary(vocabularies);
    }

    private synchronized void loadVocabularies() {
        // do what?
        allVocabularies = vocaRepo.getAllVocabulary();
    }

    public void editVocabulary(Vocabulary vocabulary) {
        vocaRepo.editVocabulary(vocabulary);
    }

    public LiveData<Vocabulary> getRandomVocabulary() {
        return vocaRepo.getRandomVocabulary();
    }

    public LiveData<Boolean> isEmpty() {
        final MutableLiveData<Boolean> result = new MutableLiveData<>(true);
        if (allVocabularies == null || allVocabularies.getValue() == null) {
            loadVocabularies();
            allVocabularies.observeForever(new Observer<List<Vocabulary>>() {
                @Override
                public void onChanged(List<Vocabulary> vocabularies) {
                    result.setValue(vocabularies.size() == 0);
                    allVocabularies.removeObserver(this);
                }
            });
        } else {
            result.setValue(allVocabularies.getValue().size() == 0);
        }
        return result;
    }
}

File: app/src/main/java/hsk/practice/myvoca/VocaViewModel.kt
package hsk.practice.myvoca

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import database.Vocabulary
import database.source.VocaRepository
import java.util.*

/**
 * VocaViewModel is at the top of the database abstraction.
 * ViewModel interacts with the UI classes directly.
 * All database operations must be done through this class.
 *
 * Also, all methods work asynchronously because database access is costly.
 * Methods return the LiveData immediately when the method is called. Actual result will be filled into LiveData later.
 * UI classes should observe the LiveData and define what to do when the operation is actually finished.
 */
class VocaViewModel : ViewModel() {

    private var allVocabularies: LiveData<MutableList<Vocabulary?>?>? = null
    private val vocabulary: Vocabulary? = null
    private val vocaRepo: VocaRepository? = VocaRepository.getInstance()

    fun getAllVocabulary(): LiveData<MutableList<Vocabulary?>?>? {
        if (allVocabularies?.value == null) {
            loadVocabularies()
        }
        return allVocabularies
    }

    fun getVocabularyCount(): LiveData<Int?> {
        val result = MutableLiveData<Int?>()
        if (allVocabularies?.value == null) {
            loadVocabularies()
            allVocabularies?.observeForever(object : Observer<MutableList<Vocabulary?>?> {
                override fun onChanged(vocabularies: MutableList<Vocabulary?>?) {
                    result.setValue(vocabularies?.size)
                    allVocabularies?.removeObserver(this)
                }
            })
        } else {
            result.setValue(allVocabularies!!.value?.size)
        }
        return result
    }

    fun deleteVocabulary(vararg vocabularies: Vocabulary?) {
        vocaRepo?.deleteVocabularies(*vocabularies)
    }

    fun getVocabulary(query: String?): LiveData<MutableList<Vocabulary?>?>? = vocaRepo?.getVocabulary(query)

    fun insertVocabulary(vararg vocabularies: Vocabulary?) = vocaRepo?.insertVocabulary(*vocabularies)

    @Synchronized
    private fun loadVocabularies() {
        // do what?
        allVocabularies = vocaRepo?.getAllVocabulary()
    }

    fun editVocabulary(vocabulary: Vocabulary?) {
        vocaRepo?.editVocabulary(vocabulary)
    }

    fun getRandomVocabulary() = vocaRepo?.getRandomVocabulary()

    fun getRandomVocabularies(count: Int, notInclude: Vocabulary?): MutableList<Vocabulary?>? {
        val result = ArrayList<Vocabulary?>()
        while (result.size < count) {
            val voca = vocaRepo?.getRandomVocabulary()?.value
            if (voca != notInclude) {
                result.add(voca)
            }
        }
        return result
    }

    fun isEmpty(): LiveData<Boolean?> {
        val result = MutableLiveData(true)
        if (allVocabularies?.value == null) {
            loadVocabularies()
            allVocabularies?.observeForever(object : Observer<MutableList<Vocabulary?>?> {
                override fun onChanged(vocabularies: MutableList<Vocabulary?>?) {
                    result.value = (vocabularies?.size == 0)
                    allVocabularies!!.removeObserver(this)
                }
            })
        } else {
            result.setValue(allVocabularies!!.value?.size == 0)
        }
        return result
    }

    init {
        loadVocabularies()
    }
}